developer/wip/for-contributors/data-plane/custom-data-plane.md

```diff
@@ -0,0 +1,252 @@
+# Writing a custom data plane
+
+<!-- TOC -->
+* [Writing a custom data plane](#writing-a-custom-data-plane)
+  * [1. The Registration Phase](#1-the-registration-phase)
+  * [2. Handling DPS messages ](#2-handling-dps-messages)
+    * [2.1 `START`](#)
+        * [2.1.1 `PUSH`](#211-push)
+        * [2.1.2 `PULL`](#212-pull) 
+    * [2.2 `SUSPEND` and `TERMINATE`](#22-suspend-and-terminate)
+
+<!-- TOC -->
+
+When not running with an embedded data plane embedded, EDC uses the Data Plane Signaling protocol ([DPS](./data-plane-signaling/data-plane-signaling.md)) for the communication between control planes and data planes. In this chapter we will see how to leverage on [DPS](./data-plane-signaling/data-plane-signaling.md) for writing a custom data plane from scratch.
+
+Just as example this chapter contains JS snippet using `express` as web framework.
+
+>  Since it's just for didactic purpose the code is not to be considered complete since proper error and JSON-LD handling are are not implemented.
+
+Our simple data plane setup looks like this:
+
+```javascript
+const express = require('express')
+const app = express()
+const port = 3000
+
+app.use(express.json());
+
+app.use((req, res, next) => {
+    console.log(req.method, req.hostname, req.path, new Date(Date.now()).toString());
+    next();
+})
+
+app.listen(port, () => {
+    console.log(`Data plane listening on port ${port}`)
+})
+
+```
+
+We start with a basic `express` application that listens on port `3000` and logs every request with a basic middleware.
+
+## 1. The Registration Phase
+
+First we need to register our custom data plane in the EDC control plane. 
+
+By using the internal `Dataplane Selector` API available under the `control` context of EDC, we could send a registration request:
+
+```http request
+POST https://controlplane-host:port/api/control/v1/dataplanes
+Content-Type: application/json
+{
+    "@context": {
+        "edc": "https://w3id.org/edc/v0.0.1/ns/"
+    },
+    "@type": "DataPlaneInstance",
+    "@id": "custom_dataplane",
+    "url": "http://custom-dataplane-host:3000/dataflows",
+    "allowedSourceTypes": [
+        "HttpData"
+    ],
+    "allowedDestTypes": [
+        "HttpData"
+    ],
+    "allowedTransferTypes": [
+        "HttpData-PULL",
+        "HttpData-PUSH"
+    ]
+}
+```
+
+> When the data plane is registered is up to the implementors. This may be a manual operation as well as an in process routine.
```

Comment:
```suggestion
> It's up to the implementors to decide when the data plane gets registered. This may be a manual operation as well as automated in a process routine.
```